### ISSUE TITLE
Tweakwise won't fetch/display sibling categories in filters

### ISSUE THREAD

**Issue by ArjenMiedema** [EmicoEcommerce/Magento2Tweakwise-archived#129]
Issue Brief
--
Showing the sibling categories is no longer working after the changes made in [this commit](https://github.com/EmicoEcommerce/Magento2Tweakwise/commit/96b777748ad53964f6fa3c34f0b951de31cb2d5c#diff-7b24e9452b88648b4b0550c82396e5413d05def3044fe42e4480d8fb6a01b09aR264). Because the parent category isn't fetched anymore when generating the tn_cid parameter for the call to Tweakwise, only child categories are loaded, meaning that on the deepest category in the tree, no categories are shown at all.

Environment
--
* PHP Version: 7.3
* Magento Version: 2.4.1
* Tweakwise Version: 3.0.2
* Tweakwise Export Version: 1.6.3
* Magento Deploy Mode: production & development
* Third party modules: N/A

Steps to reproduce
--
1. Go to the deepest category page in the tree
2. Debug the call from Magento to Tweakwise for the navigation
3. Check the results of the XML for the Tweakwise call to `navigation`

Actual result
--
1. Only `FEED` and `AFGELEID` facets are in the results
2. The `tn_cid` parameter in the URL contains only the current category (i.e. `&tn_cid=1000124`)

Expected result
--
1. There's a `CATEGORY` facet within the results
2. The `tn_cid` parameter in the URL contains the current category and the parent category (i.e. `&tn_cid=1000123-1000124`)

**Comment by edwinljacobs** [comment]
Hi,

I recall changing this behaviour for a specific reason, perhaps I made an error there. I'll look into it. If I understand correctly you are saying that only the current category is in the request and you expect the entire category path? 
Regarding: "There's a CATEGORY facet within the results" Are you saying you don't have a category facet at all?

I hope to have an update this week.

**Comment by ArjenMiedema** [comment]
@edwinljacobs when I'm on the deepest category in the tree, so with no subcategories of its own, it will not return a `CATEGORY` facet. This is because when you also have the parent category in the `tn_cid` parameter, it will check both and now it will only check the current category itself. I created a patch which combines the original functionality that was in an older version and the new functionality.

This is my patch, maybe it is useful for you as well:

```
diff --git a/src/Model/Catalog/Layer/Url/Strategy/QueryParameterStrategy.php b/src/Model/Catalog/Layer/Url/Strategy/QueryParameterStrategy.php
index 9d9cb96..7ec820b 100644
--- a/src/Model/Catalog/Layer/Url/Strategy/QueryParameterStrategy.php
+++ b/src/Model/Catalog/Layer/Url/Strategy/QueryParameterStrategy.php
@@ -21,6 +21,7 @@ use Magento\Catalog\Model\Category;
 use Magento\Framework\Exception\NoSuchEntityException;
 use Magento\Catalog\Api\CategoryRepositoryInterface;
 use Emico\TweakwiseExport\Model\Helper as ExportHelper;
+use Magento\Framework\Registry;


 class QueryParameterStrategy implements UrlInterface, FilterApplierInterface, CategoryUrlInterface
@@ -82,11 +83,13 @@ class QueryParameterStrategy implements UrlInterface, FilterApplierInterface, Ca
     public function __construct(
         UrlModel $url,
         CategoryRepositoryInterface $categoryRepository,
-        ExportHelper $exportHelper
+        ExportHelper $exportHelper,
+        Registry $registry
     ) {
         $this->url = $url;
         $this->categoryRepository = $categoryRepository;
         $this->exportHelper = $exportHelper;
+        $this->registry = $registry;
     }

     /**
@@ -273,6 +276,29 @@ class QueryParameterStrategy implements UrlInterface, FilterApplierInterface, Ca
     protected function getCategoryFilters(MagentoHttpRequest $request)
     {
         $categories = $request->getQuery(self::PARAM_CATEGORY);
+
+        if (!$categories) {
+            /** @var Category $currentCategory */
+            $currentCategory = $this->registry->registry('current_category');
+            $currentCategoryId = (int) $currentCategory->getId();
+            $parentCategoryId  = (int) $currentCategory->getParentCategory()->getId();
+
+            if (!$currentCategoryId || $currentCategoryId === 1 || !$parentCategoryId) {
+                return [];
+            }
+
+            $rootCategoryId = (int) $currentCategory->getStore()->getRootCategoryId();
+
+            if (in_array($parentCategoryId, [1, $rootCategoryId], true)) {
+                return [];
+            }
+
+            return [
+                $parentCategoryId,
+                $currentCategoryId
+            ];
+        }
+
         $categories = explode(self::CATEGORY_TREE_SEPARATOR, $categories);
         $categories = array_map('intval', $categories);
         $categories = array_filter($categories);
```

**Comment by edwinljacobs** [comment]
Thanks for the patch, I'll look into it. I hope to have an update on wednesday

With kind regards

**Comment by edwinljacobs** [comment]
Hi, I hoped to provide you with an update yesterday, that didn't work out. Ill look into this today.

**Comment by edwinljacobs** [comment]
Hi,

I was unsure where to integrate the fix the whole category filter stuff floats around in several places, I am not particularly happy about that. I opted to add a similar fix, this has been integrated in branch: issue/129-category-filter-path
With your patch applied do you have an active category filter on the category pages as in the screenshot?
![active-category-filter](https://user-images.githubusercontent.com/11278380/98238946-7a5deb00-1f67-11eb-8ebf-50c7e82597c2.png)

I decided to put a similar version of your patch in method "addCategoryFilter" as follows:

```php
/**
     * @param Category|int $category
     * @return $this
     */
    public function addCategoryFilter($category)
    {
        $ids = [];
        if (is_numeric($category)) {
            $ids[] = $category;
            return $this->addCategoryPathFilter($ids);
        }
        $ids[] = (int) $category->getId();
        /** @var Category $category */
        $parent = $category;
        while (((int) $parent->getParentId()) !== 0) {
            $parent = $parent->getParentCategory();
            $ids[] = (int) $parent->getId();
        }

        $ids = array_diff($ids, [1, (int)$category->getStore()->getRootCategoryId()]);
        $ids = array_reverse($ids);

        return $this->addCategoryPathFilter($ids);
    }
```

Im also not particularly happy with the ->getParentCategory() calls in the method above. Ideally we do not want to load those. It might be enough to just add the parent category instead of the full path.

**Comment by edwinljacobs** [comment]
I removed the parent category loading and made use of $category->getParentId(); The current category and the parent category are now added to the request when the parent categoryId is not one of 0, 1 and store root categoryId. Made sure that the category filter isnt rendered in the active filters (Layer\State block) except when we are dealing with a search requests.

If you want to verify that this works as desired (and or have any comments) feel free to do so in branch ```issue/129-category-filter-path```
If not i'll verify and release this next wednesday.

With kind regards

**Comment by ArjenMiedema** [comment]
@edwinljacobs we've done some checks and I think this should work as expected. We're currently having one shop live with this branch enabled, just to test it and there are no errors found. Once the new tag is released, we'll do some final checks and use that one instead.

**Comment by edwinljacobs** [comment]
Hi this has been released in v3.0.4: https://github.com/EmicoEcommerce/Magento2Tweakwise/releases/tag/v3.0.4

With kind regards

**Comment by ArjenMiedema** [comment]
@edwinljacobs thanks!
